core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,197 @@
+   * pass the filter, it will fail in `EB` with `None`
+   */
+  final def filter(f: B => Boolean): ZRef[EA, Option[EB], A, B] =
+    collect { case b if f(b) => b }
+
+  /**
+   * Transforms elements that are retrieved from this Ref with a function.
+   */
+  final def map[C](bc: B => C): ZRef[EA, EB, A, C] =
+    new ZRef[EA, EB, A, C] {
+      def set(a: A): IO[EA, Unit]                                             = self.set(a)
+      def get: IO[EB, C]                                                      = self.get.map(bc)
+      def setAsync(a: A): zio.IO[EA, Unit]                                    = self.setAsync(a)
+      private[zio] def unsafeModify[E, C](f: Any => (Any, Any)): zio.IO[E, C] = self.unsafeModify(f)
+      private[zio] def unsafeUpdate[E](f: Any => Any): zio.IO[E, C]           = self.unsafeUpdate(f).map(bc)
+    }
+
+  /**
+   * Transforms elements that are set into this Ref with a function.
+   */
+  final def contramap[C](ca: C => A): ZRef[EA, EB, C, B] =
+    new ZRef[EA, EB, C, B] {
+      def set(c: C): IO[EA, Unit]                                             = self.set(ca(c))
+      def get: IO[EB, B]                                                      = self.get
+      def setAsync(c: C): zio.IO[EA, Unit]                                    = self.setAsync(ca(c))
+      private[zio] def unsafeModify[E, C](f: Any => (Any, Any)): zio.IO[E, C] = self.unsafeModify(f)
+      private[zio] def unsafeUpdate[E](f: Any => Any): zio.IO[E, B]           = self.unsafeUpdate(f)
+
+    }
+
+  /**
+   * Maps from `EB` error type to another.
+   */
+  final def mapError[E](f: EB => E): ZRef[EA, E, A, B] = new ZRef[EA, E, A, B] {
+    def set(a: A): IO[EA, Unit]                                             = self.set(a)
+    def get: IO[E, B]                                                       = self.get.mapError(f)
+    def setAsync(a: A): zio.IO[EA, Unit]                                    = self.setAsync(a)
+    private[zio] def unsafeUpdate[E](f: Any => Any): zio.IO[E, B]           = self.unsafeUpdate(f)
+    private[zio] def unsafeModify[E, C](f: Any => (Any, Any)): zio.IO[E, C] = self.unsafeModify(f)
+  }
+
+  /**
+   * Maps from `EA` error type to another.
+   */
+  final def mapErrorSet[E](f: EA => E): ZRef[E, EB, A, B] = new ZRef[E, EB, A, B] {
+    def set(a: A): IO[E, Unit]                                              = self.set(a).mapError(f)
+    def get: IO[EB, B]                                                      = self.get
+    def setAsync(a: A): zio.IO[E, Unit]                                     = self.setAsync(a).mapError(f)
+    private[zio] def unsafeUpdate[E](f: Any => Any): zio.IO[E, B]           = self.unsafeUpdate(f)
+    private[zio] def unsafeModify[E, C](f: Any => (Any, Any)): zio.IO[E, C] = self.unsafeModify(f)
+  }
+
+  /**
+   * Replace the content of the Ref with `c`
+   */
+  final def const[C](c: C): ZRef[EA, EB, A, C] = map(_ => c)
```

Comment:
Be sure to sort all methods in here alphabetically. The order is:

 - abstract types (alphabetically in this category)
 - abstract operators (alphabetically in this category)
 - abstract named defs / vals  (alphabetically in this category)
 - concrete types (alphabetically in this category)
 - concrete operators (alphabetically in this category)
 - concrete named methods (alphabetically in this category)